enigma-principal/app/src/boot_network/principal_manager.rs

```diff
@@ -63,117 +117,196 @@ impl PrincipalManager {
+    fn get_block_number(&self) -> Result<U256, Error>;
 
-    fn get_account_address(&self) -> Address;
+    fn register<G: Into<U256>>(&self, signing_address: String, gas_limit: G) -> Result<H256, Error>;
 
-    fn get_network_url(&self) -> String;
+    fn verify_identity_or_register<G: Into<U256>>(&self, gas_limit: G) -> Result<Option<H256>, Error>;
 
     /// after initiation, this will run the principal node and block.
     fn run<G: Into<U256>>(&self, gas: G) -> Result<(), Error>;
 }
 
 impl Sampler for PrincipalManager {
-    fn new(config_path: &str, contract: EnigmaContract, eid: sgx_enclave_id_t) -> Result<Self, Error> {
-        let config = PrincipalManager::load_config(config_path)?;
-        Ok(Self::new_delegated(config, contract, eid))
+    fn new(config: PrincipalConfig, contract: Arc<EnigmaContract>, report_manager: ReportManager) -> Result<Self, Error> {
+        let eid = report_manager.eid;
+//        let registration_params = report_manager.get_registration_params()?;
+        Ok(PrincipalManager { config, contract, report_manager, eid })
     }
 
-    fn new_delegated(config: PrincipalConfig, contract: EnigmaContract, eid: sgx_enclave_id_t) -> Self {
-        let as_service = service::AttestationService::new(&config.attestation_service_url);
-        PrincipalManager { eid, config, as_service, contract }
-    }
+    fn get_signing_address(&self) -> Result<String, Error> { Ok(self.report_manager.get_signing_address()?) }
 
     fn get_contract_address(&self) -> Address { self.contract.address() }
 
-    fn get_quote(&self) -> Result<String, Error> { Ok(retry_quote(self.eid, &self.config.spid, 18)?) }
+    fn get_account_address(&self) -> Address { self.contract.account.clone() }
 
-    fn get_report(&self, quote: &str) -> Result<(Vec<u8>, service::ASResponse), Error> {
-        let (rlp_encoded, as_response) = self.as_service.rlp_encode_registration_params(quote)?;
-        Ok((rlp_encoded, as_response))
-    }
+    fn get_network_url(&self) -> String { self.config.url.clone() }
 
-    fn get_signing_address(&self) -> Result<String, Error> {
-        let mut signing_address = esgx::equote::get_register_signing_address(self.eid)?;
-        // remove 0x
-        signing_address = signing_address[2..].to_string();
-        Ok(signing_address)
+    fn get_block_number(&self) -> Result<U256, Error> {
+        let block_number = self.get_web3().eth().block_number().wait().unwrap();
+        Ok(block_number)
     }
 
-    fn get_account_address(&self) -> Address { self.contract.account }
+    fn register<G: Into<U256>>(&self, signing_address: String, gas_limit: G) -> Result<H256, Error> {
+        let registration_params = self.report_manager.get_registration_params()?;
+        let report = registration_params.report.clone();
+        let signature = registration_params.signature.clone();
+        let receipt = self.contract.register(signing_address, report, signature, gas_limit, self.config.confirmations as usize)?;
```

Comment:
```suggestion
        let receipt = self.contract.register(signing_address, registration_params.report, registration_params.signature, gas_limit, self.config.confirmations as usize)?;
```
You can pass the report and signature directly without cloning them in memory

Reply:
Fixed in commit: e7adf35481143153d4b637aa8f6ae93499172d2e.